compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,341 @@ fn check_for_underconstrained_values_within_function(
+                .collect(),
+            root_results: HashSet::from_iter(results.iter().copied()),
+        }
+    }
+
+    /// Add children of a given parent to the tainted value set
+    /// (for arguments one set is enough, for results we keep them
+    /// separate as the forthcoming check considers the call covered
+    /// if all the results were properly covered)
+    fn update_children(&mut self, parent: &ValueId, children: &[ValueId]) {
+        if self.arguments.contains(parent) {
+            self.arguments.extend(children);
+        }
+        for result_status in &mut self.results.iter_mut() {
+            match result_status {
+                // Skip updating results already found covered
+                ResultStatus::Constrained => {
+                    continue;
+                }
+                ResultStatus::Unconstrained { descendants } => {
+                    if descendants.contains(parent) {
+                        descendants.extend(children);
+                    }
+                }
+            }
+        }
+    }
+
+    /// If Brillig call is properly constrained by the given ids, return true
+    fn check_constrained(&self) -> bool {
+        // If every result has now been constrained,
+        // consider the call properly constrained
+        self.results.iter().all(|result| matches!(result, ResultStatus::Constrained))
+    }
+
+    /// Remember partial constraints (involving some of the results and an argument)
+    /// along the way to take them into final consideration
+    /// Generally, a valid partial constraint should link up a result descendant
+    /// and an argument descendant, although there are also edge cases mentioned below.
+    fn store_partial_constraints(&mut self, constrained_values: &HashSet<ValueId>) {
+        let mut results_involved: Vec<usize> = vec![];
+
+        // For a valid partial constraint, a value descending from
+        // one of the results should be constrained
+        for (i, result_status) in self.results.iter().enumerate() {
+            match result_status {
+                // Skip checking already covered results
+                ResultStatus::Constrained => {
+                    continue;
+                }
+                ResultStatus::Unconstrained { descendants } => {
+                    if descendants.intersection(constrained_values).next().is_some() {
+                        results_involved.push(i);
+                    }
+                }
+            }
+        }
+
+        // Along with it, one of the argument descendants should be constrained
+        // (skipped if there were no arguments, or if an actual result and not a
+        // descendant has been constrained, e.g. against a constant)
+        if !results_involved.is_empty()
+            && (self.arguments.is_empty()
+                || self.root_results.intersection(constrained_values).next().is_some()
```

Comment:
What happens in this example?

```rust
fn main(a: Field, b: Field) -> pub Field {
  let c = unsafe { add(a, b) };
  let d = unsafe { add(a, b) };
  assert_eq(c, d);
  (c + d)
}

unconstrained fn add(a: Field, b: Field) -> Field { 
  a + b
}
```

So here we have a constraint on the direct outputs of `add` but it doesn't have anything to do with its inputs. It looks like any implementation can be plugged in.

Reply:
Indeed. Fixing

Reply:
Done